crates/codex/src/ttype/atomic/object/shaped.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::BTreeMap;
+
+use serde::Deserialize;
+use serde::Serialize;
+
+use mago_atom::Atom;
+use mago_atom::atom;
+
+use crate::ttype::TType;
+use crate::ttype::TypeRef;
+use crate::ttype::atomic::array::key::ArrayKey;
+use crate::ttype::union::TUnion;
+
+/// Metadata for a PHP object (map/dictionary-like).
+///
+/// Corresponds to `object{'key': TVal, 1: TVal2 ...}` shape.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Eq, Hash, PartialOrd, Ord, Default)]
+pub struct TShapedObject {
+    /// Specific types known for certain keys (`ArrayKey`). The bool indicates if the element is optional.
+    pub known_items: Option<BTreeMap<ArrayKey, (bool, TUnion)>>,
```

Comment:
```suggestion
    pub known_items: Option<BTreeMap<Atom, (bool, TUnion)>>,
```

Property names are all strings. not array keys.